### DISCUSSION TITLE
Priorities

### DISCUSSION THREAD

**Discussion by LADSoft** [LADSoft/OrangeC#780] [Q&A]
There are a good number of things planned for the future; however; there is only so much time.   It has been indicated I may not have quite a good perception as to what is most wanted, so, I'm asking for help in setting priorities.   Please vote on things you would most like to see. The major things in the pipeline are:

finish getting the libcxx tests to compile
C++17 support
C++20 support
C2x support
x64 processor support/WIN64
ARM support
power pc support
IDE/DEBUGGER support
Better optimizations

the plan was to continue next with C++17 support once finished with the libcxx tests, however, it has been suggested that it might be more valuable to get x64 processor support done.

**Comment by chuggafan** [comment]
I would say that C++17 support should not take nearly the same amount of time as getting x64 processor support done in general, and I would actually prefer seeing work on ARM getting started beforehand compared to x64 OR PPC64 since I believe that those would iron out issues in the design of the adl files.

**Comment by GitMensch** [comment]
I'd also prefer, definitely after libcxx and possibly after posix improvements (which was part of the c++17 milestone) - processors over c*2*.
ARM64  / x64 (depending on what takes longer  / is likely to show more issues [at least x64 is likely available ot all devs]), then the other, then power, then potentially ARM32.

I personally think that a working debugger would help, also potentially in revealing issues rooted in OrangeC that are otherwise harder to inspect.

**Comment by LADSoft** [comment]
I'm not too concerned about debugging as I'm quite used to using a MAP file and correllating with assembly files at this point. 

As much as I would love to do ARM64 next, I'm thinking it might be strategically better to do x64 instead.   My relative familiarity with that platform, along with the desire to have more code generation in the adl compiler (e.g. for gen.cpp in the occ backend and some other stuff) and it might go through much more quickly than if we try an entire new architecture from scratch.     We could do x64 then do a quick trip through c++17 and c2x, then go on to ARM64 which would find bugs and incompatibilities in the newly fleshed-out adl compiler.   Meanwhile I could handle smaller stuff the way I am now, just fit it in when I need a break from the main project I'm working on.

**Comment by 1997tn** [comment]
x86_64 support, unless you wanted to be a compiler for DOS. Everything is 64 bit now!

**Comment by LADSoft** [comment]
So I've been thinking about this for several weeks in terms of what needs to be done and what I'd like to do, and what I think I'd *like* to work on is x64 support.    So that will be the next major thing I do I think...   even before I finish the libcxx testing.   Right now I'm getting close to finishing up a major rewrite to how inlining is done...   it generates much better code for iterators.   then I have a small amount of work to do on exception handling, I want to rework it slightly and fix another bug I found somewhere along the way.    Then it will be time to start working on x64...

**Comment by LADSoft** [comment]
About to wrap up backlog cleanup for now other than day to day maintenance for bugs....   we got through a lot of new features in the last couple of months...    aside from some work I'm trying to do to speed up compiles, the only new feature that I'm planning to do is to support delay loading of dlls.    After that the development path is up for grabs...

thinking some more, like @chuggafan says it will be a quicker hit to get things like the C23 and C++2017 support in than to go all out with new backends.    There is a lot to learn and beyond that there is a lot of work to be done to get the entire project compatible with 64-bit compiles in general and WIN64 in specific.

Anyway i'm currently proposing that next up is new language features and some more work with libcxx testing, and then following that we'll finally get to new backends.

@chuggafan which would you like me to work on?   C23 or C++2017?    I know you are working on one of the more complex bits of C23 but I could get some of the other stuff in?   

On the other hand I looked at the scope of work and just as a guess I think C++2017 will take significantly less time than C23 unless I wreck things badly and have to spend a lot of time resurrecting what we've got of the libcxx tests again...

**Reply by chuggafan** [reply]
I personally don't care enough, C23 and C++17 are both relatively small updates to the existing feature sets, C23 mainly outside of stuff like `embed` is optional features or library updates, with a few *small* updates elsewhere, I can go and find the exact `diff` between C18 and C23 if you so choose, I know it's one of the papers...

**Comment by LADSoft** [comment]
So the idea of portiing to other architectures keeps getting pushed off and pushed off and pushed off...  the main thing right now has been well the C++17 support needs to be further vetted, the LIBCXX tests fail in various ways...    I decided though C++17 works well enough for what I want to do right now...   ti compiles itself and it compiles at least some real-world programs I've worked with, so  given that I really want to start porting I am going to abandon c++17 temporarily and just get on with the WIN64 port.   I will get back to LIBCX17++ testing after that...   I will rewrite the milestones to reflect the new prioirities when I get a chance.

**Reply by chuggafan** [reply]
This works with me, just means that I may need to make another set of cpu-optimized compiler-generated/hand-written assembly....
I do think adding some intrinsics would work out nicely here because they would allow us to banish some of that inline assembly we generate. e.g. I know in the 32-bit code we manipulate the `fs` register a lot, needing to do that still I would say we can manipulate it through something akin to `__writefsword` or `__writegsword`.

Microsoft also does have their ABI for exceptions posted online, which is kind of shocking for me....
https://learn.microsoft.com/en-us/cpp/cpp/exception-handling-in-visual-cpp?view=msvc-170

**Comment by LADSoft** [comment]
so as far as the intriinsics go, what you wrote up seems really good...   but i don't see myself getting to it any time soon.   If you want to take that on as a long term project at some point I certainly wouldn't mind :smile:

**Reply by chuggafan** [reply]
My problem is I know the underlying solution but implementation is kind of up in the air.
My last comment in #577 was kind of me trying to think through what hypothetical IR re-write would look like, in part because my ultimate goal has always been with that "idea" is the ability to have register allocation burden moved from manually allocating registers, to allowing `adl` files to define what acceptable registers something has, and allowing that + IR-level intrinsics to allow for automatic register coloring to enable better optimizations, e.g. if we have a `bsf`, being able to not have to worry about the `fastcall` and it being able to directly be inlined into one instruction instead of needing a sequence, or if we implement `popcount`, being able to have an IR-level inline or an instruction, depending on what ultimately we use....

I think the approach that LLVM takes of having them be "functions" that can be called is a good one, the problem is my compiler theory is severely lacking and despite having worked with you on this for so long my understanding of the underbelly of `occopt` is probably the worst out of the three portions of `occ`, probably due to being scared to touch it since so much theory is required there and when I look at the code my understanding is poor, I've tried to read the papers before but man those are dense.

If I tackle this soon (if at all), I may start out with an IR-level change to support this as a building block (function call changes as mentioned), and having a real textual source <-> bitstream conversion, I think that would help a lot and the reason I requested it so long ago was because I noted how good LLVM's `opt` was for seeing the effects of individual passes. It's also good because `llvm` uses it to great effect to see how instructions get lowered into assembly and preventing regressions that way.

I'd do some level of IR-change as shown there probably in order to just deal with how we do the calls in the first place, I'm kind of partial to LLVM's style of a function call integrating the arguments as opposed to a separate IR-level paramsetup in my head due to being able to inline in one spot without needing to deal with everything. We either way will get attributes upon attributes to deal with function calling conventions etc.

Sorry if this is a bit rambly, it's half a brain dump and half a collection of thoughts that I've had for a long time about that particular side-project.

**Reply by GitMensch** [reply]
Just to point out: https://deepwiki.com/LADSoft/OrangeC (especially with source map and/or deep-thinking [that sometimes needs a lunch break!]) may be useful to have a better overview of some parts that you are less familiar with.

**Comment by LADSoft** [comment]
so yeah there is a lot of theory going on in occopt...   we've got the gcse, loop optimizations, register coloring, live variable analysis, SSA and some other stuff and each piece is somewhat complex.   I'm not sure if you should care at this point beyond basics of how to feed them code they won't entirely garble up...

let me start by saying, i like your idea about attaching the parameters to the gosub and not having separate IR instructions related to the calling convention.    We could go beyond attaching the parameters though and get rid of the stack adjustment and also associate a return temporary with the gosub.    I rather like the idea of associating a return value with it (which should be optional because of the difference between methods and functions).   This because there is a certain amount of complexity in the compiler related to tracking return values, which could later be taken out of the project if the return values are just treated like any other temporary.

I don't think the latter would be too complicated, just some detail to deal with here and there and again the streaming/unstreaming has to change....   probably the worst part of it is moving the generation of the calling convention completely to occ.exe.   A lot of the fastcall support in the compiler could be removed I suppose as well, which is another benefit, but there would have to be minor changes to the precolor algorithm so that the parameters could be precolored appropriately....   if they are precolored the register allocator will not try to modify the colors.    (for example DIV and SHIFT instructions are handled that way, and fastcall arguments are too...   there is other stuff too)

something else that may be useful eventually in this regards is to explictly add the 'thiscall' linkage type to occparse but that wouldn't have to be done as part of this effort....

I think you said there would be reasons sometimes to have code generated by occopt, but others to be generated by the backend.

When you generate code within occopt, you could generate it early on and take advantage of optimizations.   There are a couple of minor rules you need to follow when generating the code but as long as you follow them the optimzations should just accept your code and optiimize it the way compiler-generated code gets optimized....   that leaves the stuff generated by the backend.   For that, the main concern within OCCOPT at this point is that the code get marked with a fence...   later if we want to get more advanced we could remove the fence and recode any appropriate optimizations...  

you can generally generate code ad-hoc in the backed but the registers you are using need to be colored by the register allocator within occopt....    there are three caveats to that, right now the register allocator relies on an optional answer node and one or two argument nodes...   to stay within that framework only three registers can be colored by the register allocator.    We'd have to rework parts of the allocator to get past that...   another caveat is some instructions may limit the number of arguments that can have registers, or which registers can be used....    but this information is already available in the ADL file so we could generate code or data structures which describe that... the other caveat is that I could see that the backend *may* need one or more temporary registers to work with which also need to be colored somehow.   Although right now the way that is done (for example for atomics) is occopt generates a list of live registers for each instruction and then the backend can fudge something either by choosing one not in use or pushing one not used in the current context....   I never really liked that though and was hoping I could find a way to take the live data completely out of the backend lol...

**Reply by chuggafan** [reply]
> let me start by saying, i like your idea about attaching the parameters to the gosub and not having separate IR instructions related to the calling convention. We could go beyond attaching the parameters though and get rid of the stack adjustment and also associate a return temporary with the gosub. I rather like the idea of associating a return value with it (which should be optional because of the difference between methods and functions). This because there is a certain amount of complexity in the compiler related to tracking return values, which could later be taken out of the project if the return values are just treated like any other temporary.

Honestly, I'm somewhat partial to the idea of LLVM's where everything is SSA-form, all of the time, it makes some code generation headaches to be sure, but it also enables you to never have to have the headache of translating to/from SSA form for any reason, and most optimizations seem to like being done on SSA-form anyways from all of the literature I've read. If we treat everything in SSA form in the first place this may just make it easier from that perspective.

> I don't think the latter would be too complicated, just some detail to deal with here and there and again the streaming/unstreaming has to change.... probably the worst part of it is moving the generation of the calling convention completely to occ.exe. A lot of the fastcall support in the compiler could be removed I suppose as well, which is another benefit, but there would have to be minor changes to the precolor algorithm so that the parameters could be precolored appropriately.... if they are precolored the register allocator will not try to modify the colors. (for example DIV and SHIFT instructions are handled that way, and fastcall arguments are too... there is other stuff too)

Couldn't we make those entirely IR-level instructions, then have the colorization happen entirely in the backend? Especially with how common they are. I'm not (exactly) sure, but I think GCC/clang handle colorization entirely in the backend, I know LLVM has to be this way because their IR doesn't handle specific platform registers, I'm pretty sure GCC behaves this way as well. We'd also have to have individual functions be able to precolor specific inputs, so if we have an intrinsic-type function we should be able to precolor some inputs and outputs, e.g. `__writefsword` (or really any segment register) requires `EAX` as the input/output register when dealing with offsets. I think the "intrinsics as functions" will probably require a scan of all BBs in a first-pass...

> I think you said there would be reasons sometimes to have code generated by occopt, but others to be generated by the backend.

Mostly I'm thinking of how IR-level works, we'd want the IRs to actually be able to be seen in the middle-end, in order to have the optimizer be able to play around with the IR-level instructions (i.e. delete BBs, reduce `mov`s, not have to deal with prologue/epilog cleanup on the function call, etc.). This is sort-of why I was focusing on IR-level vs backend intrinsics, as my *ultimate* goal was to have the optimizer in play to not have to deal with writing fastcall intrinsics where each register is specified.

> you can generally generate code ad-hoc in the backed but the registers you are using need to be colored by the register allocator within occopt.... there are three caveats to that, right now the register allocator relies on an optional answer node and one or two argument nodes... to stay within that framework only three registers can be colored by the register allocator. 

Isn't that a pretty big issue for some SIMD instructions? I know there's plenty right now where 3 registers are simultaneously colored, but APX (which is in the new Panther Lake CPUs) has 3 registers and an immediate, and I don't doubt we'll eventually get four total registers (dest, source, source2, mask) eventually, my source for the new instructions is [this document](https://www.intel.com/content/www/us/en/content-details/836199/intel-advanced-vector-extensions-10-2-intel-avx10-2-architecture-specification.html)

> another caveat is some instructions may limit the number of arguments that can have registers, or which registers can be used.... but this information is already available in the ADL file so we could generate code or data structures which describe that.

Agreed, we need to figure out a standardized way to get that information passed from the ADL into the backend, able to be consumed for code generation.

**Comment by LADSoft** [comment]
the problem with staying in SSA all the time is that some of the optimizations take advantage of the fact that temps can be reused; the same sequence of temps means the same thing any time it occurs.    The GCSE for example is entirely dependent on that, we'd have to come up with a new algorithm if we stayed in SSA.   While it might not be a bad thing to do GCSE another way it would be a lot of effort...   Other things occopt does might be better done when working with the SSA form though....

hm i didn't think about dong the register coloring in the backend, that is possibly a good idea?    We could also completely move the address mode rewriting to the back end...   making the optimizer completely independent of the generated code sounds like a really good thing...    thing is there may be dependencies on things like the live variable analysis and so forth that would also have to be copied into the back end, because they'd be used in both modules.    But still it sound reasonable....   especially since I want to make kind of a library that would be used in common against different back ends... 

So we already do three registers (answer, left,, right) it is just that the answer and left are usually the same for x86...   i think the register allocator can probably unscramble the equality of answer and left when we get that far....   as far as adding an additional immediate i don't see much of an issue with that....    we'll deal with four-register instructions when we get to them :smile:

**Reply by chuggafan** [reply]
For GCSE, wouldn't you just need to track phis and see if the inputs to a phi compute to the same? Or am I misunderstanding something from a local CSE to global CSE changeover?
Probably need to read the code some more and understand what is being handled there...

**Comment by LADSoft** [comment]
GCSE s a little more complex than that as the PHIS only appear in blocks that have two or more predecessors...    so for example you couldn't optimize an if-else that way.....
I had AI tell me about how to do GCSE on a program in SSA form...   I get the drift of it but I'm a little foggy on the details.    But it seems like it would be way easier than what we are doing now, and much less memory intensive too lol
